From e203f92776533cbe70b750c5aef21bbf916fe186 Mon Sep 17 00:00:00 2001
From: danielnachun <daniel.nachun@gmail.com>
Date: Thu, 23 Jul 2026 12:00:00 -0700
Subject: [PATCH] port PgenReader to newer pgenlib API

The bundled pgenlib was replaced with a newer plink-ng whose reader API
changed: PgfiInitPhase1 gained a pgi_fname parameter, and the PgrGet*
functions now take a PgrSampleSubsetIndex instead of a raw cumulative
popcounts pointer. Adapt PgenReader to match, using the plink-ng
pgenlibr wrapper as the reference.
---
 submods/Pgenlib/PgenReader.cpp | 20 +++++++++++---------
 submods/Pgenlib/PgenReader.h   |  1 +
 2 files changed, 12 insertions(+), 9 deletions(-)

diff --git a/submods/Pgenlib/PgenReader.cpp b/submods/Pgenlib/PgenReader.cpp
index cfa246e..32d68ee 100644
--- a/submods/Pgenlib/PgenReader.cpp
+++ b/submods/Pgenlib/PgenReader.cpp
@@ -86,7 +86,7 @@ void PgenReader::Load(string filename, uint32_t *raw_sample_ct, uint32_t* raw_ma
     plink2::PgenHeaderCtrl header_ctrl;
     uintptr_t pgfi_alloc_cacheline_ct;
     char errstr_buf[plink2::kPglErrstrBufBlen];
-    if (PgfiInitPhase1(fname, cur_variant_ct, cur_sample_ct, 0, &header_ctrl, _info_ptr, &pgfi_alloc_cacheline_ct, errstr_buf) != plink2::kPglRetSuccess) {
+    if (PgfiInitPhase1(fname, nullptr, cur_variant_ct, cur_sample_ct, &header_ctrl, _info_ptr, &pgfi_alloc_cacheline_ct, errstr_buf) != plink2::kPglRetSuccess) {
         stop(&(errstr_buf[7]));
     }
     const uint32_t raw_variant_ct = _info_ptr->raw_variant_ct;
@@ -190,6 +190,7 @@ void PgenReader::Load(string filename, uint32_t *raw_sample_ct, uint32_t* raw_ma
 
     _subset_cumulative_popcounts = reinterpret_cast<uint32_t*>(pgr_alloc_iter);
     pgr_alloc_iter = &(pgr_alloc_iter[cumulative_popcounts_byte_ct]);
+    plink2::PgrSetSampleSubsetIndex(_subset_cumulative_popcounts, _state_ptr, &_subset_index);
     _pgv.genovec = reinterpret_cast<uintptr_t*>(pgr_alloc_iter);
     pgr_alloc_iter = &(pgr_alloc_iter[genovec_byte_ct]);
     if (multiallelic_hc_byte_ct) {
@@ -326,7 +327,7 @@ void PgenReader::ReadIntHardcalls(vector<int32_t> &buf, int variant_idx, int all
         sprintf(errstr_buf, "buf has wrong length (%" PRIdPTR "; %u expected)", buf.size(), _subset_size);
         stop(errstr_buf);
     }
-    plink2::PglErr reterr = PgrGet1(_subset_include_vec, _subset_cumulative_popcounts, _subset_size, variant_idx, allele_idx, _state_ptr, _pgv.genovec);
+    plink2::PglErr reterr = PgrGet1(_subset_include_vec, _subset_index, _subset_size, variant_idx, allele_idx, _state_ptr, _pgv.genovec);
     if (reterr != plink2::kPglRetSuccess) {
         char errstr_buf[256];
         sprintf(errstr_buf, "PgrGet1() error %d", static_cast<int>(reterr));
@@ -351,7 +352,7 @@ void PgenReader::ReadRawHard(uintptr_t *buf, int variant_idx, int allele_idx){
         stop("buffer size not enough");
     }
     */
-    plink2::PglErr reterr = PgrGet1(_subset_include_vec, _subset_cumulative_popcounts, _subset_size, variant_idx, allele_idx, _state_ptr, buf);
+    plink2::PglErr reterr = PgrGet1(_subset_include_vec, _subset_index, _subset_size, variant_idx, allele_idx, _state_ptr, buf);
     if (reterr != plink2::kPglRetSuccess) {
         char errstr_buf[256];
         sprintf(errstr_buf, "PgrGet1() error %d", static_cast<int>(reterr));
@@ -374,7 +375,7 @@ void PgenReader::ReadRawFullHard(uintptr_t *buf, int variant_idx, int allele_idx
     }
     */
     //cout << "start read" << std::endl;
-    plink2::PglErr reterr = PgrGet1(NULL, NULL, _info_ptr->raw_sample_ct, variant_idx, allele_idx, _state_ptr, buf);
+    plink2::PglErr reterr = PgrGet1(NULL, _subset_index, _info_ptr->raw_sample_ct, variant_idx, allele_idx, _state_ptr, buf);
     /* // not much faster
     const unsigned char* fread_ptr;
     const unsigned char* fread_end;
@@ -399,7 +400,7 @@ void PgenReader::ReadRawFullHard(uintptr_t *buf, int variant_idx){
         sprintf(errstr_buf, "variant_num out of range (%d; must be 1..%u)", variant_idx + 1, _info_ptr->raw_variant_ct);
         stop(errstr_buf);
     }
-    plink2::PglErr reterr = PgrGet(NULL, NULL, _info_ptr->raw_sample_ct, variant_idx,  _state_ptr, buf);
+    plink2::PglErr reterr = PgrGet(NULL, _subset_index, _info_ptr->raw_sample_ct, variant_idx,  _state_ptr, buf);
 
     if (reterr != plink2::kPglRetSuccess) {
         char errstr_buf[256];
@@ -580,7 +581,7 @@ void PgenReader::ReadHardcalls(vector<double> &buf, int variant_idx, int allele_
         sprintf(errstr_buf, "buf has wrong length (%" PRIdPTR "; %u expected)", buf.size(), _subset_size);
         stop(errstr_buf);
     }
-    plink2::PglErr reterr = PgrGet1(_subset_include_vec, _subset_cumulative_popcounts, _subset_size, variant_idx, allele_idx, _state_ptr, _pgv.genovec);
+    plink2::PglErr reterr = PgrGet1(_subset_include_vec, _subset_index, _subset_size, variant_idx, allele_idx, _state_ptr, _pgv.genovec);
     if (reterr != plink2::kPglRetSuccess) {
         char errstr_buf[256];
         sprintf(errstr_buf, "PgrGet1() error %d", static_cast<int>(reterr));
@@ -604,7 +605,7 @@ void PgenReader::Read(vector<double> &buf, int variant_idx, int allele_idx) {
         stop(errstr_buf);
     }
     uint32_t dosage_ct;
-    plink2::PglErr reterr = PgrGet1D(_subset_include_vec, _subset_cumulative_popcounts, _subset_size, variant_idx, allele_idx, _state_ptr, _pgv.genovec, _pgv.dosage_present, _pgv.dosage_main, &dosage_ct);
+    plink2::PglErr reterr = PgrGet1D(_subset_include_vec, _subset_index, _subset_size, variant_idx, allele_idx, _state_ptr, _pgv.genovec, _pgv.dosage_present, _pgv.dosage_main, &dosage_ct);
     if (reterr != plink2::kPglRetSuccess) {
         char errstr_buf[256];
         sprintf(errstr_buf, "PgrGet1D() error %d", static_cast<int>(reterr));
@@ -623,7 +624,7 @@ void PgenReader::ReadDosage(uintptr_t *buf, int variant_idx, int allele_idx){
     //std::cout << "genovec_size64: " << genovec_size64 << ", dosage present: " << dosage_present_size64 << ", dosage main: " << dosage_main_size64 << std::endl;
     //plink2::PglErr reterr = PgrGet1D(_subset_include_vec, _subset_cumulative_popcounts, _subset_size, variant_idx, allele_idx, _state_ptr, genovec, dosage_present, dosage_main, dosage_ct_ptr);
     uint32_t dosage_ct;
-    plink2::PglErr reterr = PgrGet1D(_subset_include_vec, _subset_cumulative_popcounts, _subset_size, variant_idx, allele_idx, _state_ptr, genovec, dosage_present, dosage_main, dosage_ct_ptr);
+    plink2::PglErr reterr = PgrGet1D(_subset_include_vec, _subset_index, _subset_size, variant_idx, allele_idx, _state_ptr, genovec, dosage_present, dosage_main, dosage_ct_ptr);
     if(reterr != plink2::kPglRetSuccess){
         char errstr_buf[256];
         sprintf(errstr_buf, "PgrGet1D() error %d", static_cast<int>(reterr));
@@ -962,6 +963,7 @@ void PgenReader::SetSampleSubsetInternal(const vector<uint32_t> &sample_subset_0
 
     const uint32_t raw_sample_ctl = plink2::DivUp(raw_sample_ct, plink2::kBitsPerWord);
     plink2::FillCumulativePopcounts(_subset_include_vec, raw_sample_ctl, _subset_cumulative_popcounts);
+    plink2::PgrSetSampleSubsetIndex(_subset_cumulative_popcounts, _state_ptr, &_subset_index);
     _subset_size = sample_subset_0based.size();
     _genovecbuf_uptr_size = plink2::DivUp(_subset_size, plink2::kNypsPerVec) * plink2::kBytesPerVec;
 }
@@ -972,7 +974,7 @@ void PgenReader::ReadAllelesPhasedInternal(int variant_idx) {
         sprintf(errstr_buf, "variant_num out of range (%d; must be 1..%u)", variant_idx + 1, _info_ptr->raw_variant_ct);
         stop(errstr_buf);
     }
-    plink2::PglErr reterr = plink2::PgrGetMP(_subset_include_vec, _subset_cumulative_popcounts, _subset_size, variant_idx, _state_ptr, &_pgv);
+    plink2::PglErr reterr = plink2::PgrGetMP(_subset_include_vec, _subset_index, _subset_size, variant_idx, _state_ptr, &_pgv);
     if (reterr != plink2::kPglRetSuccess) {
         char errstr_buf[256];
         sprintf(errstr_buf, "PgrGetMP() error %d", static_cast<int>(reterr));
diff --git a/submods/Pgenlib/PgenReader.h b/submods/Pgenlib/PgenReader.h
index ad3e194..a175552 100644
--- a/submods/Pgenlib/PgenReader.h
+++ b/submods/Pgenlib/PgenReader.h
@@ -146,6 +146,7 @@ class PgenReader {
         uintptr_t* _subset_include_vec;
         uintptr_t* _subset_include_interleaved_vec;
         uint32_t* _subset_cumulative_popcounts;
+        plink2::PgrSampleSubsetIndex _subset_index;
         uint32_t _subset_size;
         uint32_t _raw_sample_ct;
         
-- 
2.53.0

